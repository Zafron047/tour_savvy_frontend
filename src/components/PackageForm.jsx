import React, { useState } from 'react';
import { useDispatch } from 'react-redux';
import { createPackage, fetchPackages } from '../redux/packages/packagesSlice';
import '../stylesheets/package_form.css';

const PackageForm = () => {
  const [name, setName] = useState('');
  const [description, setDescription] = useState('');
  const [image, setImage] = useState('');
  const [packageTypes, setPackageTypes] = useState([{ name: '', price: '', description: '' }]);
  const dispatch = useDispatch();
  const addPackageType = () => {
    setPackageTypes([...packageTypes, { name: '', price: '', description: '' }]);
  };
  const handlePackageTypeChange = (index, e) => {
    const updatedPackageTypes = [...packageTypes];
    updatedPackageTypes[index][e.target.name] = e.target.value;
    setPackageTypes(updatedPackageTypes);
  };
  const handleSubmit = async (e) => {
    e.preventDefault();
    const newPackage = {
      name,
      description,
      image,
      package_type: packageTypes,
    };
    await dispatch(createPackage(newPackage));
    await dispatch(fetchPackages());
    setName('');
    setDescription('');
    setImage('');
    setPackageTypes([{ name: '', price: '', description: '' }]);
  };
  const user = JSON.parse(localStorage.getItem('user'));
  if (!user) {
    return <div>Please log in to add a package.</div>;
  }
  return (
    <div className="PackageForm">
      <form className="input" onSubmit={handleSubmit}>
        <div className="mb-3">
          <input type="text" placeholder="Name" id="nameInput" className="form-control" value={name} onChange={(e) => setName(e.target.value)} required />
        </div>
        <div className="mb-3">
          <textarea className="form-control" placeholder="Description" value={description} onChange={(e) => setDescription(e.target.value)} required />
        </div>
        <div className="mb-3">
          <input type="text" placeholder="Image URL" className="form-control" value={image} onChange={(e) => setImage(e.target.value)} required />
        </div>
        {packageTypes.map((packageType, index) => (
          /* eslint-disable react/no-array-index-key */
          <div key={index} className="mb-4">
            <div className="mb-3">
              <input
                type="text"
                name="name"
                placeholder="Package"
                className="form-control"
                value={packageType.name}
                onChange={(e) => handlePackageTypeChange(index, e)}
                required
              />
            </div>
            <div className="mb-3">
              <input
                type="number"
                name="price"
                placeholder="Price"
                className="form-control"
                value={packageType.price}
                onChange={(e) => handlePackageTypeChange(index, e)}
                required
              />
            </div>
            <div className="mb-3">
              <textarea
                name="description"
                placeholder="Package Plan"
                className="form-control"
                value={packageType.description}
                onChange={(e) => handlePackageTypeChange(index, e)}
                required
              />
            </div>
          </div>
        ))}
        <div className="pkg-btn-wrap">
          <button type="button" className="btn btn-secondary" onClick={addPackageType}>
            Add Package
          </button>
          <button type="submit" className="btn btn-primary create">Create Package</button>
        </div>
      </form>
    </div>
  );
};
export default PackageForm;
