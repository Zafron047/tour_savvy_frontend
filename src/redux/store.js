import { configureStore } from '@reduxjs/toolkit';
import packagesSlice from './packages/packagesSlice';
import reservationsReducer from './reservations/reservationSlice';
import userSlice from './authentication/userSlice';

const loadStateFromLocalStorage = () => {
  try {
    const serializedState = localStorage.getItem('reduxState');
    if (serializedState === null) {
      return undefined;
    }
    return JSON.parse(serializedState);
  } catch (err) {
    return undefined;
  }
};

const loadUserFromLocalStorage = () => {
  try {
    const serializedUser = localStorage.getItem('user');
    if (serializedUser === null) {
      return undefined;
    }
    return JSON.parse(serializedUser);
  } catch (err) {
    return undefined;
  }
};

const store = configureStore({
  reducer: {
    packages: packagesSlice,
    reservations: reservationsReducer,
    current_user: userSlice,
  },
  preloadedState: {
    current_user: loadUserFromLocalStorage(),
    ...loadStateFromLocalStorage(),
  },
  reservations: reservationsReducer,
});

export default store;
