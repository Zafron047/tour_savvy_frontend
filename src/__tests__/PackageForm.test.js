import React from 'react';
import { render } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';
import { Provider } from 'react-redux';
import store from '../redux/store';
import PackageForm from '../components/PackageForm';

test('Should render PackageForm component correctly', () => {
  const { container } = render(
    <Provider store={store}>
      <BrowserRouter>
        <PackageForm />
      </BrowserRouter>
    </Provider>,
  );
  expect(container).toMatchSnapshot();
});
