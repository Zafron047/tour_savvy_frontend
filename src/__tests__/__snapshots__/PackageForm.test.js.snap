// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render PackageForm component correctly 1`] = `
<div>
  <div>
    Please log in to add a package.
  </div>
</div>
`;
